File: packages/trustix/api/log.go
// Copyright (C) 2021 Tweag IO
//
// This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, version 3.
//
// This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.

package api

// Common log operations

import (
	"context"
	"fmt"

	"github.com/tweag/trustix/packages/trustix-proto/api"
	"github.com/tweag/trustix/packages/trustix-proto/schema"
	vlog "github.com/tweag/trustix/packages/trustix/log"
	"github.com/tweag/trustix/packages/trustix/storage"
)

func getLogConsistencyProof(prefix string, txn storage.Transaction, ctx context.Context, req *api.GetLogConsistencyProofRequest) (resp *api.ProofResponse, err error) {
	resp = &api.ProofResponse{}

	vLog, err := vlog.NewVerifiableLog(prefix, txn, *req.SecondSize)
	if err != nil {
		return nil, err
	}

	proof, err := vLog.ConsistencyProof(*req.FirstSize, *req.SecondSize)
	if err != nil {
		return nil, err
	}

	resp.Proof = proof

	return resp, nil
}

func getLogAuditProof(prefix string, txn storage.Transaction, ctx context.Context, req *api.GetLogAuditProofRequest) (resp *api.ProofResponse, err error) {

	vLog, err := vlog.NewVerifiableLog(prefix, txn, *req.TreeSize)
	if err != nil {
		return nil, err
	}

	proof, err := vLog.AuditProof(*req.Index, *req.TreeSize)
	if err != nil {
		return nil, err
	}

	resp.Proof = proof

	return resp, nil
}

func getLogEntries(prefix string, txn storage.Transaction, ctx context.Context, req *api.GetLogEntriesRequest) (*api.LogEntriesResponse, error) {

	resp := &api.LogEntriesResponse{
		Leaves: []*schema.LogLeaf{},
	}

	logStorage := vlog.NewLogStorage(prefix, txn)

	for i := int(*req.Start); i <= int(*req.Finish); i++ {
		leaf, err := logStorage.Get(0, uint64(i))
		if err != nil {
			return nil, err
		}
		resp.Leaves = append(resp.Leaves, leaf)
	}

	total := *req.Finish - *req.Start + 1
	if uint64(len(resp.Leaves)) != total {
		return nil, fmt.Errorf("%d != %d", len(resp.Leaves), total)
	}

	return resp, nil
}

File: packages/trustix-nix-reprod/trustix_nix_reprod/models/log.py
# Trustix
# Copyright (C) 2021 Tweag IO

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

from tortoise import models
from tortoise import fields

from trustix_nix_reprod.models.meta import BaseMeta


class Log(models.Model):
    name = fields.CharField(
        max_length=55,
        index=True,
    )
    tree_size = fields.IntField()

    def __str__(self):
        return self.name

    class Meta(BaseMeta):
        pass
